### PR TITLE
[meshcop] implement R bit of Security Policy TLV

### PR DESCRIPTION
**Pull request by jwhui**

### CODE REVIEW

File: src/core/thread/mle_router.cpp [hunk 1]
```diff
@@ -99,14 +99,21 @@ void MleRouter::HandlePartitionChange(void)
...
 }
 
-void MleRouter::SetRouterRoleEnabled(bool aEnabled)
+otError MleRouter::SetRouterEligible(bool aEnabled)
 {
-    mRouterRoleEnabled = aEnabled;
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(IsFullThreadDevice() &&
+                     (GetNetif().GetKeyManager().GetSecurityPolicyFlags() & OT_SECURITY_POLICY_ROUTERS),
```
Comment (Nopik) [inline src/core/thread/mle_router.cpp]: I can't shake the feeling that this `VerifyOrExit` will occasionally jump over `SetBeaconEnabled` and/or `BecomeDetached` calls below. I'm thinking about the complex corner case like: device starts, `mRouterEligible` is `true` from constructor, and then somebody calls `SetRouterEligible( false )` which before would call `BecomeDetached()` and now it wont. Maybe that is a problem, maybe not (after all, device would need to become leader or router first) - though it is far from obvious to me that this code is fully correct.

Like, is it possible for Security Policy TLV to change bit after devices are commissioned? Existing leader & routers would need to drop their roles, I don't see a code like that.

Comment (jwhui) [reply src/core/thread/mle_router.cpp]: Thanks for raising this issue.  I added some logic in `Mle::HandleStateChanged()` to address this issue.  Please review.

Comment (Nopik) [reply src/core/thread/mle_router.cpp]: LGTM

**Comment by jwhui** [comment]
Putting this PR on hold since it may cause Thread Test 5.8.4 to fail.  In particular, the test involves performing a `MGMT_ACTIVE_SET.req` to update the `R` flag.  However, the test cannot complete since the Leader immediately becomes an end device as a result.  I believe the test (and possibly spec) need to be updated to indicate that changing the `R` flag should be considered a breaking change and done via the Pending Operational Dataset.  We need to raise this discussion within the Thread Group.

**Comment by jwhui** [comment]
This issue is being discussed in Thread Group JIRA [SPEC-768](https://threadgroup.atlassian.net/browse/SPEC-768).

**Comment by jwhui** [comment]
Superseded by #9187

**Pull request closed without merging**